From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Zurker <91011862+hershate@users.noreply.github.com>
Date: Fri, 31 Jul 2026 23:23:49 +0800
Subject: [PATCH] ResultSlot ItemCraftedEvent 2 onQuickCraft onTake Ca


diff --git a/net/minecraft/world/inventory/ResultSlot.java b/net/minecraft/world/inventory/ResultSlot.java
index 32e2ee436578e8f35d1cbadf3eee31a8bee1404b..b6036e405b713f307797dcddc74f423956fedf5c 100644
--- a/net/minecraft/world/inventory/ResultSlot.java
+++ b/net/minecraft/world/inventory/ResultSlot.java
@@ -40,7 +40,7 @@ public class ResultSlot extends Slot {
         this.removeCount += amount;
         this.checkTakeAchievements(stack);
         // Paper start - add ItemCraftedEvent
-        if (!stack.isEmpty()) {
+        if (!stack.isEmpty() && io.papermc.paper.event.inventory.ItemCraftedEvent.getHandlerList().getRegisteredListeners().length > 0) { // Papo - zero-listener gate
             new io.papermc.paper.event.inventory.ItemCraftedEvent((org.bukkit.entity.Player) this.player.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(stack)).callEvent();
         }
         // Paper end - add ItemCraftedEvent
@@ -86,7 +86,7 @@ public class ResultSlot extends Slot {
     @Override
     public void onTake(Player player, ItemStack stack) {
         // Paper start - add ItemCraftedEvent
-        if (!stack.isEmpty()) {
+        if (!stack.isEmpty() && io.papermc.paper.event.inventory.ItemCraftedEvent.getHandlerList().getRegisteredListeners().length > 0) { // Papo - zero-listener gate
             new io.papermc.paper.event.inventory.ItemCraftedEvent((org.bukkit.entity.Player) player.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(stack)).callEvent();
         }
         // Paper end - add ItemCraftedEvent
